src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp

```diff
@@ -726,6 +741,65 @@ ConcreteCommandPath CodegenDataModelProvider::NextGeneratedCommand(const Concret
     return ConcreteCommandPath(before.mEndpointId, before.mClusterId, *commandId);
 }
 
+std::optional<DataModel::DeviceTypeEntry> CodegenDataModelProvider::FirstDeviceType(EndpointId endpoint)
+{
+    std::optional<unsigned> endpoint_index = TryFindEndpointIndex(endpoint);
+    if (!endpoint_index.has_value())
+    {
+        return std::nullopt;
+    }
+
+    CHIP_ERROR err                            = CHIP_NO_ERROR;
+    Span<const EmberAfDeviceType> deviceTypes = emberAfDeviceTypeListFromEndpointIndex(*endpoint_index, err);
```

Comment:
So why did we need to add `emberAfDeviceTypeListFromEndpointIndex`?  We return nullopt both on not having an index and on having an empty list... and `emberAfDeviceTypeListFromEndpoint` just returns an empty list if it's not a valid endpoint, no?

Reply:
The reason was that the codegen data provider has a cache for finding the endpoint index, so it is more efficient in that case. I was trying to avoid having the iterator do a `findEndpoint` lookup on every Next call, as it is less efficient as-is already and did not want to move it to O(n^2) on number of endpoints.

Reply:
Specifically I thought the call of `emberAfIndexFromEndpoint(endpoint)` was unnecessarely expensive to be called in a Next loop.

Reply:
I placed this in FirstDeviceType for consistency. Added a comment.